python-code/code.py

```
@@ -0,0 +1,125 @@
+#the FIPS data does not come with column names
+census_col_names = ["STATE","STATEFP","COUNTYFP","COUNTYNAME","CLASSFP"]
+
+# Open url, read in the data with the column names, and convert specific columns to str.
+# When Pandas reads these columns, it automatilcally intrepets them as INTS
+fips_df = pd.read_table(
+    census_fips_url,
+    sep=",",
+    names=census_col_names,
+    converters={'STATEFP': str,'COUNTYFP': str,'CLASSFP': str}
+)
+
+# Combine State & County FP to generate the full FIPS code for easier lookup
+fips_df["STCOUNTYFP"] = fips_df["STATEFP"] + fips_df["COUNTYFP"]
+#Dropping STATFP & COUNTYFP as we no longer need them
+fips_df = fips_df[["STCOUNTYFP", "STATE" ,"COUNTYNAME", "CLASSFP"]]
+
+# Get current year to handle future runs of this file
+now = dt.datetime.now()
+cur_year = now.year
+
+def get_files_url(month, year):
+    monthyear = month + str(year)
+    return "https://www.huduser.gov/portal/datasets/usps/ZIP_COUNTY_{}.xlsx".format(monthyear)
+
+"""
+Main loop
+"""
+def main()
+    # from the beginning of hud year data to current year
+    for year in range(2010, cur_year+1):
+        #hud files are based on quarters
+        for month in ["03", "06", "09", "12"]:
+            #generate the HUDs url
+            url = get_files_url(month, year)
+            #download the file
+            full_file_path = download_file(url)
+            #open and get the excel dataframe
+            excel_df = process_excel_file(full_file_path)
+            #merge the excel file with the fips data
+            merged_df = fips_df.merge(excel_df)
+            #reduce the dataframe down to specific columns
+            merged_df = merged_df[["ZIP", "COUNTYNAME", "STATE", "STCOUNTYFP", "CLASSFP"]]
+            #generate a csv file path
+            csv_path = get_file_path(full_file_path, True)
+            print(csv_path)
+            try:
+                merged_df.to_csv(csv_path, encoding='utf-8', index=False)
+            except:
```

This `except` is too broad. When I first tried to run this, the code did nothing because it was trying to write to a directory that didn't exist. It didn't show me the error because this clause swallowed it. If you're expecting a particular error, you should catch it as specifically as possible.